kms/uri/uri.go

```diff
@@ -0,0 +1,86 @@
+package uri
+
+import (
+	"net/url"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+// URI implements a parser for a URI format based on the the PKCS #11 URI Scheme
+// defined in https://tools.ietf.org/html/rfc7512
+//
+// These URIs will be used to define the key names in a KMS.
+type URI struct {
+	*url.URL
+	Values url.Values
+}
+
+// New creates a new URI from a scheme and key-value pairs.
+func New(scheme string, values url.Values) *URI {
+	return &URI{
+		URL: &url.URL{
+			Scheme: scheme,
+			Opaque: strings.ReplaceAll(values.Encode(), "&", ";"),
```

Comment:
How do we know values don't have `;` in them?

Reply:
It _might_ be safe considering there's an old rfc that allows `;` as a query separator.

Reply:
A `;` is a reserverd character and will be %-encoded, the `Encode()` method will encode using `&` as a separator, we're replacing it with `;` because the RFC that this package is based-on uses `;` as a separator.